### PR TITLE
add replaceBeforeVersion field for channel

### PR DESCRIPTION
**Pull request by Ewocker**

### CODE REVIEW

File: channels/pkg/api/channel.go [hunk 1]
```diff
@@ -62,6 +62,14 @@ type AddonSpec struct {
 	// version of the software we are packaging.  But we always want to reinstall when we
 	// switch kubernetes versions.
 	Id string `json:"id,omitempty"`
+
+	// ReplaceBeforeVersion is an optional value which can be used to force channel to update
+	// addons using kubectl replace instead of kubectl apply. For example, we made a change on an
+	// immutable field in version "1.1.0-kops.2" of an addon spec which cannot be successfully updated
+	// with kubectl apply. Setting ReplaceBeforeVersion to "1.1.0-kops.2" makes sure when updating
+	// from any version below "1.1.0-kops.2", channel should update the addon using kubectl replace
```
Comment (johngmyers) [inline channels/pkg/api/channel.go]:
```suggestion
	// from any version below "1.1.0-kops.2" the channel should update the addon using kubectl replace
```

File: channels/pkg/api/channel.go [hunk 2]
```diff
@@ -62,6 +62,9 @@ type AddonSpec struct {
 	// version of the software we are packaging.  But we always want to reinstall when we
 	// switch kubernetes versions.
 	Id string `json:"id,omitempty"`
+
+	// ReplaceBeforeVersion is an optional value which can be used to ...
```
[the remark this answers is not in the document]
Comment (Ewocker) [reply channels/pkg/api/channel.go]: done

File: channels/pkg/channels/addon.go [hunk 3]
```diff
@@ -58,7 +59,8 @@ func (m *AddonMenu) MergeAddons(o *AddonMenu) {
 		if existing == nil {
 			m.Addons[k] = v
 		} else {
-			if existing.ChannelVersion().replaces(v.ChannelVersion()) {
+			replace, _ := existing.ChannelVersion().replaces(v.ChannelVersion())
+			if replace {
```
[the remark this answers is not in the document]
Comment (Ewocker) [reply channels/pkg/channels/addon.go]: done

File: channels/pkg/channels/addons_test.go [hunk 4]
```diff
@@ -84,90 +84,129 @@ func Test_Filtering(t *testing.T) {
...
 	grid := []struct {
+		Name     string
 		Old      *ChannelVersion
 		New      *ChannelVersion
 		Replaces bool
+		Force    bool
 	}{
 		// With no id, update if and only if newer semver
 		{
+			Name:     "no id - 1",
```
[the remark this answers is not in the document]
Comment (Ewocker) [reply channels/pkg/channels/addons_test.go]: done

File: channels/pkg/channels/addon.go [hunk 5]
```diff
@@ -138,7 +143,12 @@ func (a *Addon) EnsureUpdated(k8sClient kubernetes.Interface) (*AddonUpdate, err
 	}
 	klog.Infof("Applying update from %q", manifestURL)
 
-	err = Apply(manifestURL.String())
+	if required.Replace {
+		klog.Infof("Addon update requires replace")
```
[the remark this answers is not in the document]
Comment (Ewocker) [reply channels/pkg/channels/addon.go]: done

File: channels/pkg/channels/addon.go [hunk 6]
```diff
@@ -97,14 +100,16 @@ func (a *Addon) GetRequiredUpdates(k8sClient kubernetes.Interface) (*AddonUpdate
 		return nil, err
 	}
 
-	if existingVersion != nil && !newVersion.replaces(existingVersion) {
+	replace, force := newVersion.replaces(existingVersion)
```
[the remark this answers is not in the document]
Comment (Ewocker) [reply channels/pkg/channels/addon.go]: done

File: channels/pkg/channels/apply.go [hunk 7]
```diff
@@ -31,30 +31,56 @@ import (
...
-	if err != nil {
-		return fmt.Errorf("error creating temp dir: %v", err)
-	}
+	_, err = execKubectl("apply", "-f", localManifestFile)
+	return err
+}
 
+// Replace calls kubectl replace to replace the manifest. This can be a disruptive call because all resources will be deleted and recreate.
+// Note that if a resource does not previously exist, it will return an error. See "kubectl replace --help" for detailed behaviors.
+// We will likely in future change this to create things directly (or more likely embed this logic into kubectl itself)
```
[the remark this answers is not in the document]
Comment (Ewocker) [reply channels/pkg/channels/apply.go]: done

File: channels/pkg/channels/channel_version.go [hunk 8]
```diff
@@ -101,36 +102,51 @@ func (c *Channel) AnnotationName() string {
 	return AnnotationPrefix + c.Name
 }
 
-func (c *ChannelVersion) replaces(existing *ChannelVersion) bool {
+// replaces returns 2 bool. First bool is true if new ChannelVersion if higher than the existing ChannelVersion.
+// Second bool is true if new ChannelVersion should force replace the existing ChannelVersion.
```
[the remark this answers is not in the document]
Comment (Ewocker) [reply channels/pkg/channels/channel_version.go]: done

File: channels/pkg/api/channel.go (hunk 1 above)
Comment (Ewocker) [reply channels/pkg/api/channel.go]: done

**Comment by rifelpet** [comment]
It looks like protokube is crashing in the e2e test:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/8722/pull-kops-e2e-kubernetes-aws/1237985818350456833/artifacts/13.211.154.224/protokube.log

```
I0312 06:29:36.080358    9623 s3fs.go:219] Reading file "s3://k8s-kops-prow/e2e-3106cff154-ff1eb.test-cncf-aws.k8s.io/addons/bootstrap-channel.yaml"
I0312 06:29:36.660577    9623 addons.go:129] Skipping version range "<1.12.0" that does not match current version 1.19.0
I0312 06:29:36.660602    9623 addons.go:129] Skipping version range "<1.12.0" that does not match current version 1.19.0
I0312 06:29:36.660610    9623 addons.go:129] Skipping version range "<1.15.0" that does not match current version 1.19.0
I0312 06:29:36.662650    9623 channel_version.go:108] Checking existing channel: <nil> compared to new channel: Version=1.5.0 Channel=s3://k8s-kops-prow/e2e-3106cff154-ff1eb.test-cncf-aws.k8s.io/addons/bootstrap-channel.yaml ManifestHash=2ea50e23f1a5aa41df3724630ac25173738cc90c
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x16ef913]
goroutine 1 [running]:
k8s.io/kops/channels/pkg/channels.(*ChannelVersion).updates(0xc0007c1300, 0x0, 0xc000458b40)
	channels/pkg/channels/channel_version.go:109 +0x93
k8s.io/kops/channels/pkg/channels.(*Addon).GetRequiredUpdates(0xc00043cf20, 0x20bc6e0, 0xc000458b40, 0x0, 0x0, 0x0)
	channels/pkg/channels/addon.go:100 +0x17f
k8s.io/kops/channels/pkg/cmd.RunApplyChannel(0x203ef60, 0xc000449010, 0x2041060, 0xc0000aa008, 0xc000462de0, 0xc000474600, 0x1, 0x3, 0x0, 0x0)
	channels/pkg/cmd/apply_channel.go:136 +0x609
k8s.io/kops/channels/pkg/cmd.NewCmdApplyChannel.func1(0xc00047c000, 0xc000474600, 0x1, 0x3, 0x0, 0x0)
	channels/pkg/cmd/apply_channel.go:44 +0x70
k8s.io/kops/vendor/github.com/spf13/cobra.(*Command).execute(0xc00047c000, 0xc000474570, 0x3, 0x3, 0xc00047c000, 0xc000474570)
	vendor/github.com/spf13/cobra/command.go:826 +0x460
k8s.io/kops/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc000465900, 0x2fc90e0, 0x0, 0x0)
	vendor/github.com/spf13/cobra/command.go:914 +0x2fb
k8s.io/kops/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	vendor/github.com/spf13/cobra/command.go:864
k8s.io/kops/channels/pkg/cmd.Execute(0x203ef60, 0xc000449010, 0x2041060, 0xc0000aa008, 0xc000371f50, 0x40576f)
	channels/pkg/cmd/root.go:40 +0x1ac
main.main()
	channels/cmd/channels/main.go:31 +0x66
```

**Comment by Ewocker** [comment]
@rifelpet Thanks.
Is there anyway I can check the namespace status of the e2e test because the nil pointer does not seem likely to be a cause related to my change (although it may be). I am not too sure about the e2e test flow, but when using the dev-build.sh, I saw multiple errors in kubelet where it says ```is forbidden: User "kubelet" cannot get resource```. Not sure if it can be related.

File: channels/pkg/channels/addon.go [hunk 9]
```diff
@@ -97,14 +97,16 @@ func (a *Addon) GetRequiredUpdates(k8sClient kubernetes.Interface) (*AddonUpdate
 		return nil, err
 	}
 
-	if existingVersion != nil && !newVersion.replaces(existingVersion) {
+	update, replace := newVersion.updates(existingVersion)
```
Comment (rifelpet) [inline channels/pkg/channels/addon.go]: It looks like `GetInstalledVersion` can return a nil `existingVersion` which I'm guessing is the case for brand new clusters that are having their channels applied for the very first time. If `existingVersion` is nil that would trigger a panic on line 112 of `channel_version.go`

Comment (Ewocker) [reply channels/pkg/channels/addon.go]: I see thanks.

**Comment by johngmyers** [comment]
Per Kops Office Hours (though taken from memory) @justinsb to do some checking with other sections of the Kubernetes organization.

**Comment by olemarkus** [comment]
What is the status of this one now?

File: go.mod [hunk 10]
```diff
@@ -53,7 +53,6 @@ require (
 	github.com/MakeNowJust/heredoc v0.0.0-20170808103936-bb23615498cd
 	github.com/Masterminds/semver v1.3.1 // indirect
 	github.com/Masterminds/sprig v2.17.1+incompatible
-	github.com/Microsoft/go-winio v0.4.14 // indirect
```
Comment (rifelpet) [inline go.mod:56]: can you remove these go.* changes? the rest of the code isn't changing any imports so I'm not sure why these would need to be removed. They're also causing a git conflict so removing the changes is probably the easiest resolution.

**Comment by rifelpet** [comment]
@olemarkus I think we're waiting on a review from @justinsb 

once this lands we should consider using it with #7919 if needed.

/assign @justinsb

**Pull request closed without merging**